src/openassetio-core/include/openassetio/hostApi/Manager.hpp

```diff
@@ -1558,6 +1558,202 @@ class OPENASSETIO_CORE_EXPORT Manager final {
+   * but there is no available default entity reference, then the
+   * `optional` entity reference will not contain a value.
+   *
+   * @throws errors.NotImplementedException Thrown when this method is
+   * not implemented by the manager. Check that this method is
+   * implemented before use by calling @ref hasCapability with @ref
+   * Capability.kDefaultEntityReferences.
+   *
+   * @see @ref Capability.kDefaultEntityReferences
+   */
+  std::variant<errors::BatchElementError, std::optional<EntityReference>> defaultEntityReference(
+      const trait::TraitSet& traitSet, access::DefaultEntityAccess defaultEntityAccess,
+      const ContextConstPtr& context, const BatchElementErrorPolicyTag::Variant& errorPolicyTag);
+
+  /**
+   * Called to determine an @ref EntityReference considered to be a
+   * sensible default for each of the given entity @ref trait_set
+   * "trait sets" and @ref Context "context".
+   *
+   * See documentation for the <!--
+   * --> @ref defaultEntityReference(const trait::TraitSets&, <!--
+   * --> access::DefaultEntityAccess, const ContextConstPtr&, <!--
+   * --> const DefaultEntityReferenceSuccessCallback&, <!--
+   * --> const BatchElementErrorCallback&)
+   * "callback signature" for more details on default entity behaviour.
+   *
+   * Any errors that occur during the query will be immediately thrown
+   * as an exception, either from the @ref manager plugin (for errors
+   * not specific to the default entity query) or as a
+   * @fqref{errors.BatchElementException}
+   * "BatchElementException"-derived error.
+   *
+   * @param traitSets The relevant trait sets for each type of entity
+   * required. These will be interpreted in conjunction with the @p
+   * context to determine the most sensible default.
+   *
+   * @param defaultEntityAccess Intended usage of the returned entity
+   * reference.
+   *
+   * @param context The calling context.
+   *
+   * @param errorPolicyTag  Parameter for selecting the appropriate
+   * overload (tagged dispatch idiom). See @ref
+   * BatchElementErrorPolicyTag::Exception.
+   *
+   * @return A list of default entity references, each corresponding an
+   * entry in @p traitSets. If the query is well-formed, but there is no
+   * available default entity reference, then the `optional` entity
+   * reference will not contain a value.
+   *
+   * @throws errors.NotImplementedException Thrown when this method is
+   * not implemented by the manager. Check that this method is
+   * implemented before use by calling @ref hasCapability with @ref
+   * Capability.kDefaultEntityReferences.
+   *
+   * @see @ref Capability.kDefaultEntityReferences
+   */
+  std::vector<std::optional<EntityReference>> defaultEntityReference(
+      const trait::TraitSets& traitSets, access::DefaultEntityAccess defaultEntityAccess,
+      const ContextConstPtr& context,
+      const BatchElementErrorPolicyTag::Exception& errorPolicyTag = {});
```

Comment:
I'm concerned with overload resolution with this, as apart from the return type, it looks the same signature as the first function above:

```
  std::optional<EntityReference> defaultEntityReference(
      const trait::TraitSet& traitSet, access::DefaultEntityAccess defaultEntityAccess,
      const ContextConstPtr& context,
      const BatchElementErrorPolicyTag::Exception& errorPolicyTag = {});
```

What were your thoughts on how to resolve between the two?

Reply:
Ah I think you missed the distinction between `BatchElementErrorPolicyTag::Exception` and `BatchElementErrorPolicyTag::Variant` - we're using the tagged dispatch idiom to allow two overloads that would otherwise only vary by their return type. In practice, if you don't specify a tag you'll get the `BatchElementErrorPolicyTag::Exception` because of the default value (`= {}`). So you have to really want the `std::variant`-returning signature and explicitly pass the `Variant` tag.

Reply:
I had seen there were two different types in the last argument. But there are pairs of scalar vs vector return, in which the signatures look the same to me. If I group them together below in their pairs, the signatures look the same for the first two (except return type), and the last two.



```
  std::optional<EntityReference> defaultEntityReference(
      const trait::TraitSet& traitSet, access::DefaultEntityAccess defaultEntityAccess,
      const ContextConstPtr& context,
      const BatchElementErrorPolicyTag::Exception& errorPolicyTag = {});

  std::vector<std::optional<EntityReference>> defaultEntityReference(
      const trait::TraitSets& traitSets, access::DefaultEntityAccess defaultEntityAccess,
      const ContextConstPtr& context,
      const BatchElementErrorPolicyTag::Exception& errorPolicyTag = {});

  std::variant<errors::BatchElementError, std::optional<EntityReference>>
  defaultEntityReference(const trait::TraitSet& traitSet,
                         access::DefaultEntityAccess defaultEntityAccess,
                         const ContextConstPtr& context,
                         const BatchElementErrorPolicyTag::Variant& errorPolicyTag);

  std::vector<std::variant<errors::BatchElementError, std::optional<EntityReference>>>
  defaultEntityReference(const trait::TraitSets& traitSets,
                         access::DefaultEntityAccess defaultEntityAccess,
                         const ContextConstPtr& context,
                         const BatchElementErrorPolicyTag::Variant& errorPolicyTag);

````

Reply:
Ahh wait, I've seen a plural! `trait::TraitSet` vs `trait::TraitSets`

That's hard to see!

Reply:
You can call my comment an 'off by one' mistake ;)

Reply:
Ah yeah, that's a common pattern and is my fault. There's also e.g. `EntityReference` vs `EntityReferences`. Interesting point that this may hurt readability.